src/types/qwlayershellv1.cpp

```diff
@@ -0,0 +1,238 @@
+    if (!handle)
+        return nullptr;
+    return new QWLayerShellV1(handle, true);
+}
+#else
+QWLayerShellV1 *QWLayerShellV1::create(QWDisplay *display)
+{
+    auto handle = wlr_layer_shell_v1_create(display->handle());
+    if (!handle)
+        return nullptr;
+    return new QWLayerShellV1(handle, true);
+}
+#endif // WLR_VERSION_MINOR > 16
+
+QWLayerShellV1 *QWLayerShellV1::get(wlr_layer_shell_v1 *handle)
+{
+    return QWLayerShellV1Private::map.value(handle);
+}
+
+QWLayerShellV1 *QWLayerShellV1::from(wlr_layer_shell_v1 *handle)
+{
+    if (auto o = get(handle))
+        return o;
+    return new QWLayerShellV1(handle, false);
+}
+
+/// QWLayerSurfaceV1
+
+class QWLayerSurfaceV1Private : public QWObjectPrivate
+{
+public:
+    QWLayerSurfaceV1Private(wlr_layer_surface_v1 *handle, bool isOwner, QWLayerSurfaceV1 *qq)
+        : QWObjectPrivate(handle, isOwner, qq)
+    {
+        Q_ASSERT(!map.contains(handle));
+        map.insert(handle, qq);
+        sc.connect(&handle->events.map, this, &QWLayerSurfaceV1Private::on_map);
+        sc.connect(&handle->events.unmap, this, &QWLayerSurfaceV1Private::on_unmap);
+        sc.connect(&handle->events.new_popup, this, &QWLayerSurfaceV1Private::on_new_popup);
+        sc.connect(&handle->events.destroy, this, &QWLayerSurfaceV1Private::on_destroy);
+    }
+    ~QWLayerSurfaceV1Private() {
+        if (!m_handle)
+            return;
+        destroy();
+        if (isHandleOwner)
+            wlr_layer_surface_v1_destroy(q_func()->handle());
+    }
+
+    inline void destroy() {
+        Q_ASSERT(m_handle);
+        Q_ASSERT(map.contains(m_handle));
+        Q_EMIT q_func()->beforeDestroy(q_func());
+        map.remove(m_handle);
+        sc.invalidate();
+    }
+
+    void on_map(void *);
+    void on_unmap(void *);
+    void on_new_popup(void *);
+    void on_destroy(void *);
+
+    static QHash<void*, QWLayerSurfaceV1*> map;
+    QW_DECLARE_PUBLIC(QWLayerSurfaceV1)
+    QWSignalConnector sc;
+};
+QHash<void*, QWLayerSurfaceV1*> QWLayerSurfaceV1Private::map;
+
+void QWLayerSurfaceV1Private::on_destroy(void *)
+{
+    destroy();
+    m_handle = nullptr;
+    delete q_func();
+}
+
+void QWLayerSurfaceV1Private::on_map(void *data)
+{
+    Q_ASSERT(m_handle == data);
+    Q_EMIT q_func()->map();
+}
+
+void QWLayerSurfaceV1Private::on_unmap(void *data)
+{
+    Q_ASSERT(m_handle == data);
+    Q_EMIT q_func()->unmap();
+}
+
+void QWLayerSurfaceV1Private::on_new_popup(void *data)
+{
+    Q_ASSERT(m_handle == data);
+    Q_EMIT q_func()->newPopup();
+}
+
+QWLayerSurfaceV1::QWLayerSurfaceV1(wlr_layer_surface_v1 *handle, bool isOwner)
+    : QObject(nullptr)
+    , QWObject(*new QWLayerSurfaceV1Private(handle, isOwner, this))
+{
+
+
+}
+
+QWLayerSurfaceV1 *QWLayerSurfaceV1::get(wlr_layer_surface_v1 *handle)
+{
+    return QWLayerSurfaceV1Private::map.value(handle);
+}
+
+QWLayerSurfaceV1 *QWLayerSurfaceV1::from(wlr_layer_surface_v1 *handle)
+{
+    if (handle == nullptr)
```

Comment:
We don't needs check the handle value in there, it's value should check from caller.